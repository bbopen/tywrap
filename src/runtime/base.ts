/**
 * Base runtime bridge
 */

import type { RuntimeExecution } from '../types/index.js';

export abstract class RuntimeBridge implements RuntimeExecution {
  abstract call<T = unknown>(
    module: string,
    functionName: string,
    args: unknown[],
    kwargs?: Record<string, unknown>
  ): Promise<T>;

  abstract instantiate<T = unknown>(
    module: string,
    className: string,
    args: unknown[],
    kwargs?: Record<string, unknown>
  ): Promise<T>;

  abstract dispose(): Promise<void>;
}
