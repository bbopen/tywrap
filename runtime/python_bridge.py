#!/usr/bin/env python3
import sys
import json
import importlib
import os
import traceback

instances = {}

FALLBACK_JSON = os.environ.get('TYWRAP_CODEC_FALLBACK', '').lower() == 'json'


def serialize(obj):
    try:
        import numpy as np  # noqa: F401
        if hasattr(obj, 'tolist') and hasattr(obj, 'shape'):
            try:
                import pyarrow as pa  # type: ignore
                import base64
                # numpy array -> Arrow tensor (fallback to tolist when not supported)
                try:
                    arr = pa.array(obj)
                    sink = pa.BufferOutputStream()
                    with pa.ipc.new_stream(sink, arr.type) as writer:
                        writer.write(arr)
                    buf = sink.getvalue()
                    b64 = base64.b64encode(buf.to_pybytes()).decode('ascii')
                    return { '__tywrap__': 'ndarray', 'encoding': 'arrow', 'b64': b64, 'shape': getattr(obj, 'shape', None) }
                except Exception:
                    if not FALLBACK_JSON:
                        raise
            except Exception:
                if not FALLBACK_JSON:
                    raise
            if FALLBACK_JSON:
                return { '__tywrap__': 'ndarray', 'encoding': 'json', 'data': obj.tolist(), 'shape': getattr(obj, 'shape', None) }
            raise RuntimeError('Arrow encoding unavailable for ndarray; install pyarrow or set TYWRAP_CODEC_FALLBACK=json to enable JSON fallback')
    except Exception:
        pass
    try:
        import pandas as pd  # noqa: F401
        # Pandas DataFrame
        if hasattr(obj, 'to_dict') and getattr(type(obj), '__name__', '') == 'DataFrame':
            try:
                import pyarrow as pa  # type: ignore
                import pyarrow.feather as feather  # type: ignore
                import base64
                try:
                    table = pa.Table.from_pandas(obj)  # type: ignore
                    sink = pa.BufferOutputStream()
                    feather.write_feather(table, sink)
                    buf = sink.getvalue()
                    b64 = base64.b64encode(buf.to_pybytes()).decode('ascii')
                    return { '__tywrap__': 'dataframe', 'encoding': 'arrow', 'b64': b64 }
                except Exception:
                    if not FALLBACK_JSON:
                        raise
            except Exception:
                if not FALLBACK_JSON:
                    raise
            if FALLBACK_JSON:
                return { '__tywrap__': 'dataframe', 'encoding': 'json', 'data': obj.to_dict(orient='records') }
            raise RuntimeError('Arrow encoding unavailable for pandas.DataFrame; install pyarrow or set TYWRAP_CODEC_FALLBACK=json to enable JSON fallback')
        # Pandas Series
        if getattr(type(obj), '__name__', '') == 'Series':
            try:
                import pyarrow as pa  # type: ignore
                import base64
                try:
                    arr = pa.Array.from_pandas(obj)  # type: ignore
                    sink = pa.BufferOutputStream()
                    with pa.ipc.new_stream(sink, arr.type) as writer:
                        writer.write(arr)
                    buf = sink.getvalue()
                    b64 = base64.b64encode(buf.to_pybytes()).decode('ascii')
                    return { '__tywrap__': 'series', 'encoding': 'arrow', 'b64': b64, 'name': getattr(obj, 'name', None) }
                except Exception:
                    if not FALLBACK_JSON:
                        raise
            except Exception:
                if not FALLBACK_JSON:
                    raise
            try:
                data = obj.to_list()  # type: ignore
            except Exception:
                data = obj.to_dict()  # type: ignore
            if FALLBACK_JSON:
                return { '__tywrap__': 'series', 'encoding': 'json', 'data': data, 'name': getattr(obj, 'name', None) }
            raise RuntimeError('Arrow encoding unavailable for pandas.Series; install pyarrow or set TYWRAP_CODEC_FALLBACK=json to enable JSON fallback')
    except Exception:
        pass
    return obj

def handle_call(params):
    module_name = params.get('module')
    function_name = params.get('functionName')
    args = params.get('args') or []
    kwargs = params.get('kwargs') or {}
    mod = importlib.import_module(module_name)
    func = getattr(mod, function_name)
    res = func(*args, **kwargs)
    return serialize(res)

def handle_instantiate(params):
    module_name = params.get('module')
    class_name = params.get('className')
    args = params.get('args') or []
    kwargs = params.get('kwargs') or {}
    mod = importlib.import_module(module_name)
    cls = getattr(mod, class_name)
    obj = cls(*args, **kwargs)
    handle_id = str(id(obj))
    instances[handle_id] = obj
    return handle_id

def main():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            msg = json.loads(line)
            mid = msg.get('id')
            method = msg.get('method')
            params = msg.get('params') or {}
            try:
                if method == 'call':
                    result = handle_call(params)
                elif method == 'instantiate':
                    result = handle_instantiate(params)
                else:
                    raise ValueError('Unknown method')
                out = { 'id': mid, 'result': result }
            except Exception as e:
                out = {
                    'id': mid,
                    'error': { 'type': type(e).__name__, 'message': str(e), 'traceback': traceback.format_exc() }
                }
        except Exception as e:
            out = { 'id': -1, 'error': { 'type': type(e).__name__, 'message': str(e) } }
        sys.stdout.write(json.dumps(out) + '\n')
        sys.stdout.flush()

if __name__ == '__main__':
    main()


