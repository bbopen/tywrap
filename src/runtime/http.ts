/**
 * HTTP runtime bridge
 */

import { RuntimeBridge } from './base.js';

export interface HttpBridgeOptions {
  baseURL: string;
  headers?: Record<string, string>;
  timeoutMs?: number;
}

interface HttpCallPayload {
  module: string;
  functionName: string;
  args: unknown[];
  kwargs?: Record<string, unknown>;
}

interface HttpInstantiatePayload {
  module: string;
  className: string;
  args: unknown[];
  kwargs?: Record<string, unknown>;
}

export class HttpBridge extends RuntimeBridge {
  private readonly baseURL: string;
  private readonly headers: Record<string, string>;
  private readonly timeoutMs: number;

  constructor(options: HttpBridgeOptions = { baseURL: 'http://localhost:8000' }) {
    super();
    this.baseURL = options.baseURL.replace(/\/$/, '');
    this.headers = { 'content-type': 'application/json', ...(options.headers ?? {}) };
    this.timeoutMs = options.timeoutMs ?? 30000;
  }

  async call<T = unknown>(
    module: string,
    functionName: string,
    args: unknown[],
    kwargs?: Record<string, unknown>
  ): Promise<T> {
    const payload: HttpCallPayload = { module, functionName, args, kwargs };
    const res = await this.post(`${this.baseURL}/call`, payload);
    return res as T;
  }

  async instantiate<T = unknown>(
    module: string,
    className: string,
    args: unknown[],
    kwargs?: Record<string, unknown>
  ): Promise<T> {
    const payload: HttpInstantiatePayload = { module, className, args, kwargs };
    const res = await this.post(`${this.baseURL}/instantiate`, payload);
    return res as T;
  }

  async dispose(): Promise<void> {
    // stateless
  }

  private async post(url: string, body: unknown): Promise<unknown> {
    const controller = typeof AbortController !== 'undefined' ? new AbortController() : undefined;
    const timer = controller ? setTimeout(() => controller.abort(), this.timeoutMs) : undefined;
    try {
      const resp = await fetch(url, {
        method: 'POST',
        headers: this.headers,
        body: JSON.stringify(body),
        signal: controller?.signal,
      });
      if (!resp.ok) {
        const text = await safeText(resp);
        throw new Error(`HTTP ${resp.status}: ${text || resp.statusText}`);
      }
      const ct = resp.headers.get('content-type') ?? '';
      if (ct.includes('application/json')) {
        return (await resp.json()) as unknown;
      }
      const text = await resp.text();
      try {
        return JSON.parse(text) as unknown;
      } catch {
        return text as unknown;
      }
    } finally {
      if (timer) {
        clearTimeout(timer);
      }
    }
  }
}

async function safeText(resp: Response): Promise<string> {
  try {
    return await resp.text();
  } catch {
    return '';
  }
}
