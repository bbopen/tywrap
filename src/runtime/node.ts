/**
 * Node.js runtime bridge (minimal MVP)
 */

import { decodeValueAsync } from '../utils/codec.js';

import { RuntimeBridge } from './base.js';

interface RpcRequest {
  id: number;
  method: 'call' | 'instantiate';
  params: unknown;
}

interface RpcResponse<T = unknown> {
  id: number;
  result?: T;
  error?: { type: string; message: string; traceback?: string };
}

export interface NodeBridgeOptions {
  pythonPath?: string;
  scriptPath?: string; // path to python_bridge.py
  cwd?: string;
  timeoutMs?: number;
  /**
   * When true, sets TYWRAP_CODEC_FALLBACK=json for the Python process to prefer JSON encoding
   * for rich types (ndarray/dataframe/series). Default: false for fast-fail on Arrow path issues.
   */
  enableJsonFallback?: boolean;
  /**
   * Optional extra environment variables to pass to the Python subprocess.
   */
  env?: Record<string, string | undefined>;
}

export class NodeBridge extends RuntimeBridge {
  private child?: {
    stdin?: NodeJS.WritableStream;
    stdout?: NodeJS.ReadableStream;
    stderr?: NodeJS.ReadableStream;
    kill: (signal?: NodeJS.Signals | number) => void;
    on: (event: 'exit' | 'error', listener: (...args: unknown[]) => void) => void;
  };
  private nextId = 1;
  private readonly pending = new Map<
    number,
    { resolve: (v: unknown) => void; reject: (e: unknown) => void; timer?: NodeJS.Timeout }
  >();
  private readonly options: Required<NodeBridgeOptions>;
  private stderrBuffer = '';

  constructor(options: NodeBridgeOptions = {}) {
    super();
    this.options = {
      pythonPath: options.pythonPath ?? 'python3',
      scriptPath: options.scriptPath ?? 'runtime/python_bridge.py',
      cwd: options.cwd ?? process.cwd(),
      timeoutMs: options.timeoutMs ?? 30000,
      enableJsonFallback: options.enableJsonFallback ?? false,
      env: options.env ?? {},
    };
  }

  async init(): Promise<void> {
    if (this.child) {
      return;
    }
    const { spawn } = await import('child_process');
    const env: NodeJS.ProcessEnv = { ...process.env, ...this.options.env };
    // Respect explicit request for JSON fallback only; otherwise fast-fail by default
    if (this.options.enableJsonFallback && !env.TYWRAP_CODEC_FALLBACK) {
      env.TYWRAP_CODEC_FALLBACK = 'json';
    }

    this.child = spawn(this.options.pythonPath, [this.options.scriptPath], {
      cwd: this.options.cwd,
      stdio: ['pipe', 'pipe', 'pipe'],
      env,
    });

    let buffer = '';
    this.child.stdout?.on('data', (chunk: Buffer): void => {
      buffer += chunk.toString();
      let idx: number;
      while ((idx = buffer.indexOf('\n')) !== -1) {
        const line = buffer.slice(0, idx);
        buffer = buffer.slice(idx + 1);
        if (!line.trim()) {
          continue;
        }
        (async (): Promise<void> => {
          try {
            const msg = JSON.parse(line) as RpcResponse;
            const pending = this.pending.get(msg.id);
            if (pending) {
              this.pending.delete(msg.id);
              if (pending.timer) {
                clearTimeout(pending.timer);
              }
              if (msg.error) {
                pending.reject(this.errorFrom(msg.error));
              } else {
                const decoded = await decodeValueAsync(msg.result);
                pending.resolve(decoded);
              }
            }
          } catch {
            // ignore invalid lines
          }
        })().catch(() => {
          /* ignore */
        });
      }
    });

    this.child.stderr?.on('data', (chunk: Buffer) => {
      // Buffer stderr for better error diagnostics on failures/exits
      try {
        this.stderrBuffer += chunk.toString();
        // Truncate to last 8KB to avoid unbounded growth
        const MAX = 8 * 1024;
        if (this.stderrBuffer.length > MAX) {
          this.stderrBuffer = this.stderrBuffer.slice(this.stderrBuffer.length - MAX);
        }
      } catch {
        // ignore
      }
    });

    this.child.on('exit', () => {
      for (const [, p] of this.pending) {
        const stderrTail = this.stderrBuffer.trim();
        const msg = stderrTail
          ? `Python process exited. Stderr:\n${stderrTail}`
          : 'Python process exited';
        p.reject(new Error(msg));
      }
      this.pending.clear();
      this.child = undefined;
    });
  }

  async call<T = unknown>(
    module: string,
    functionName: string,
    args: unknown[],
    kwargs?: Record<string, unknown>
  ): Promise<T> {
    await this.init();
    return this.send<T>({ method: 'call', params: { module, functionName, args, kwargs } });
  }

  async instantiate<T = unknown>(
    module: string,
    className: string,
    args: unknown[],
    kwargs?: Record<string, unknown>
  ): Promise<T> {
    await this.init();
    return this.send<T>({ method: 'instantiate', params: { module, className, args, kwargs } });
  }

  async dispose(): Promise<void> {
    if (!this.child) {
      return;
    }
    this.child.kill('SIGTERM');
    this.child = undefined;
  }

  private async send<T>(payload: Omit<RpcRequest, 'id'>): Promise<T> {
    const id = this.nextId++;
    const message: RpcRequest = { id, ...payload } as RpcRequest;
    const text = `${JSON.stringify(message)}\n`;
    const promise = new Promise<T>((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        const stderrTail = this.stderrBuffer.trim();
        const msg = stderrTail
          ? `Python call timed out. Recent stderr from Python:\n${stderrTail}`
          : 'Python call timed out';
        reject(new Error(msg));
      }, this.options.timeoutMs);
      const resolveWrapped = (v: unknown): void => {
        resolve(v as T);
      };
      this.pending.set(id, { resolve: resolveWrapped, reject, timer });
    });
    this.child?.stdin?.write(text);
    return promise;
  }

  private errorFrom(err: { type: string; message: string; traceback?: string }): Error {
    const e = new Error(`${err.type}: ${err.message}`);
    (e as Error & { traceback?: string }).traceback = err.traceback;
    return e;
  }
}
