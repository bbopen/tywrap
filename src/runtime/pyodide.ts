/**
 * Pyodide runtime bridge (browser)
 */

import { RuntimeBridge } from './base.js';

export interface PyodideBridgeOptions {
  indexURL: string;
  packages?: string[];
}

type LoadPyodide = (options: { indexURL: string }) => Promise<PyodideInstance>;

interface PyodideInstance {
  runPython: (code: string) => unknown;
  runPythonAsync: (code: string) => Promise<unknown>;
  globals: { get: (key: string) => unknown; set: (k: string, v: unknown) => void };
  toPy: (obj: unknown) => unknown;
  loadPackage: (name: string | string[]) => Promise<void>;
}

export class PyodideBridge extends RuntimeBridge {
  private readonly indexURL: string;
  private readonly packages: readonly string[];
  private py?: PyodideInstance;
  private initPromise?: Promise<void>;

  constructor(options: PyodideBridgeOptions = { indexURL: 'https://cdn.jsdelivr.net/pyodide/' }) {
    super();
    this.indexURL = options.indexURL;
    this.packages = [...(options.packages ?? [])];
  }

  private async ensureReady(): Promise<void> {
    if (this.py) {
      return;
    }
    // If already initializing, wait for that promise
    if (this.initPromise) {
      return this.initPromise;
    }
    // Start initialization and store the promise to prevent concurrent initialization
    this.initPromise = this.doInit();
    return this.initPromise;
  }

  private async doInit(): Promise<void> {
    const loadPyodideFn: LoadPyodide | undefined = await this.resolveLoadPyodide();
    if (!loadPyodideFn) {
      throw new Error('Pyodide is not available in this environment');
    }
    this.py = await loadPyodideFn({ indexURL: this.indexURL });
    if (this.packages.length > 0) {
      await this.py.loadPackage([...this.packages]);
    }
    await this.bootstrapHelpers();
  }

  private async resolveLoadPyodide(): Promise<LoadPyodide | undefined> {
    // Prefer global loadPyodide when present (browser script tag include)
    const g = (globalThis as unknown as { loadPyodide?: LoadPyodide }) ?? {};
    if (typeof g.loadPyodide === 'function') {
      return g.loadPyodide;
    }
    try {
      // Attempt dynamic import if available
      const mod = (await import('pyodide')) as unknown as { loadPyodide?: LoadPyodide };
      if (typeof mod.loadPyodide === 'function') {
        return mod.loadPyodide;
      }
    } catch (err) {
      // Ignore import errors - pyodide may not be installed
      // This is expected in most environments
    }
    return undefined;
  }

  private async bootstrapHelpers(): Promise<void> {
    if (!this.py) {
      return;
    }
    const helper = [
      'import importlib',
      'def __tywrap_call(module, function_name, args, kwargs):',
      '    mod = importlib.import_module(module)',
      '    fn = getattr(mod, function_name)',
      '    return fn(*args, **(kwargs or {}))',
      'def __tywrap_instantiate(module, class_name, args, kwargs):',
      '    mod = importlib.import_module(module)',
      '    cls = getattr(mod, class_name)',
      '    return cls(*args, **(kwargs or {}))',
    ].join('\n');
    await this.py.runPythonAsync(helper);
  }

  async call<T = unknown>(
    module: string,
    functionName: string,
    args: unknown[],
    kwargs?: Record<string, unknown>
  ): Promise<T> {
    await this.ensureReady();
    const py = this.py;
    if (!py) {
      throw new Error('Pyodide not initialized');
    }
    const fn = py.globals.get('__tywrap_call');
    if (!fn) {
      throw new Error('Pyodide helper not initialized');
    }
    const invoke = fn as (module: string, f: string, a: unknown, k: unknown) => unknown;
    const out = invoke(module, functionName, py.toPy(args ?? []), py.toPy(kwargs ?? {}));
    return out as T;
  }

  async instantiate<T = unknown>(
    module: string,
    className: string,
    args: unknown[],
    kwargs?: Record<string, unknown>
  ): Promise<T> {
    await this.ensureReady();
    const py = this.py;
    if (!py) {
      throw new Error('Pyodide not initialized');
    }
    const fn = py.globals.get('__tywrap_instantiate');
    if (!fn) {
      throw new Error('Pyodide helper not initialized');
    }
    const invoke = fn as (module: string, c: string, a: unknown, k: unknown) => unknown;
    const out = invoke(module, className, py.toPy(args ?? []), py.toPy(kwargs ?? {}));
    return out as T;
  }

  async dispose(): Promise<void> {
    // Pyodide has no explicit dispose for instance; rely on GC
    this.py = undefined;
  }
}
