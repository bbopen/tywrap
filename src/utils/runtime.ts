/**
 * Runtime detection utilities for Node.js, Deno, Bun, and Browser environments
 */

/// <reference path="../types/global.d.ts" />

export type Runtime = 'node' | 'deno' | 'bun' | 'browser' | 'unknown';

interface RuntimeInfo {
  name: Runtime;
  version?: string;
  capabilities: RuntimeCapabilities;
}

interface RuntimeCapabilities {
  filesystem: boolean;
  subprocess: boolean;
  webassembly: boolean;
  webworkers: boolean;
  sharedArrayBuffer: boolean;
  fetch: boolean;
}

// Cache for runtime detection to avoid repeated environment checks
let runtimeCache: RuntimeInfo | null = null;

/**
 * Clear runtime cache (for testing purposes only)
 * @internal
 */
export function clearRuntimeCache(): void {
  runtimeCache = null;
}

/**
 * Detect the current JavaScript runtime environment
 * Results are cached and frozen to prevent external mutation
 */
export function detectRuntime(): RuntimeInfo {
  if (runtimeCache) {
    return runtimeCache;
  }
  // Deno detection (must come before Node.js check)
  if (typeof Deno !== 'undefined' && Deno !== null) {
    const capabilities = {
      filesystem: true,
      subprocess: true,
      webassembly: true,
      webworkers: true,
      sharedArrayBuffer: typeof SharedArrayBuffer !== 'undefined',
      fetch: true,
    };
    const result: RuntimeInfo = {
      name: 'deno',
      version: Deno.version?.deno,
      capabilities: Object.freeze(capabilities),
    };
    runtimeCache = Object.freeze(result) as RuntimeInfo;
    return runtimeCache;
  }

  // Bun detection
  if (typeof Bun !== 'undefined' && Bun !== null) {
    const capabilities = {
      filesystem: true,
      subprocess: true,
      webassembly: true,
      webworkers: true,
      sharedArrayBuffer: typeof SharedArrayBuffer !== 'undefined',
      fetch: true,
    };
    const result: RuntimeInfo = {
      name: 'bun',
      version: Bun.version,
      capabilities: Object.freeze(capabilities),
    };
    runtimeCache = Object.freeze(result) as RuntimeInfo;
    return runtimeCache;
  }

  // Node.js detection
  if (typeof process !== 'undefined' && process.versions?.node) {
    const capabilities = {
      filesystem: true,
      subprocess: true,
      webassembly: typeof WebAssembly !== 'undefined',
      webworkers: false, // Node.js worker_threads are different
      sharedArrayBuffer: typeof SharedArrayBuffer !== 'undefined',
      fetch: typeof fetch !== 'undefined', // Available in Node.js 18+
    };
    const result: RuntimeInfo = {
      name: 'node',
      version: process.versions.node,
      capabilities: Object.freeze(capabilities),
    };
    runtimeCache = Object.freeze(result) as RuntimeInfo;
    return runtimeCache;
  }

  // Browser detection
  if (typeof window !== 'undefined' || typeof self !== 'undefined') {
    const isSecureContext =
      typeof window !== 'undefined'
        ? window.isSecureContext
        : typeof self !== 'undefined'
          ? self.isSecureContext
          : false;

    const capabilities = {
      filesystem: false,
      subprocess: false,
      webassembly: typeof WebAssembly !== 'undefined',
      webworkers: typeof Worker !== 'undefined',
      sharedArrayBuffer: isSecureContext && typeof SharedArrayBuffer !== 'undefined',
      fetch: typeof fetch !== 'undefined',
    };
    const result: RuntimeInfo = {
      name: 'browser',
      capabilities: Object.freeze(capabilities),
    };
    runtimeCache = Object.freeze(result) as RuntimeInfo;
    return runtimeCache;
  }

  const capabilities = {
    filesystem: false,
    subprocess: false,
    webassembly: false,
    webworkers: false,
    sharedArrayBuffer: false,
    fetch: false,
  };
  const result: RuntimeInfo = {
    name: 'unknown',
    capabilities: Object.freeze(capabilities),
  };

  // Cache and freeze the result to prevent external mutation
  runtimeCache = Object.freeze(result) as RuntimeInfo;
  return runtimeCache;
}

/**
 * Check if running in Node.js
 */
export function isNodejs(): boolean {
  return detectRuntime().name === 'node';
}

/**
 * Check if running in Deno
 */
export function isDeno(): boolean {
  return detectRuntime().name === 'deno';
}

/**
 * Check if running in Bun
 */
export function isBun(): boolean {
  return detectRuntime().name === 'bun';
}

/**
 * Check if running in browser
 */
export function isBrowser(): boolean {
  return detectRuntime().name === 'browser';
}

/**
 * Get runtime capabilities
 */
export function getRuntimeCapabilities(): RuntimeCapabilities {
  return detectRuntime().capabilities;
}

/**
 * Check if a specific capability is supported
 */
export function hasCapability(capability: keyof RuntimeCapabilities): boolean {
  // eslint-disable-next-line security/detect-object-injection
  return getRuntimeCapabilities()[capability];
}

/**
 * Get the best runtime strategy for Python execution
 */
export function getBestPythonRuntime(): 'node' | 'pyodide' | 'http' {
  const runtime = detectRuntime();

  if (runtime.name === 'browser') {
    return 'pyodide';
  }

  if (runtime.capabilities.subprocess) {
    return 'node'; // Works for Node.js, Deno, and Bun
  }

  // Fallback to HTTP bridge
  return 'http';
}

// Cache for lazy-loaded path module
let pathModule: any = null;

/**
 * Lazy load Node.js path module on demand
 */
async function loadPathModule(): Promise<any> {
  if (pathModule) {
    return pathModule;
  }

  const runtime = detectRuntime();
  if (runtime.name === 'node') {
    try {
      pathModule = await import('node:path');
      return pathModule;
    } catch {
      // Fallback for older Node.js versions
      pathModule = await import('path');
      return pathModule;
    }
  }
  
  return null;
}

/**
 * Normalize path by stripping '.' and resolving '..' components
 */
function normalizePath(path: string): string {
  const isAbsolute = path.startsWith('/');
  const segments = path.split('/');
  const normalized: string[] = [];
  
  for (const segment of segments) {
    if (segment === '.' || segment === '') {
      continue; // Skip current directory and empty segments
    } else if (segment === '..') {
      if (normalized.length > 0 && normalized[normalized.length - 1] !== '..') {
        normalized.pop(); // Go up one directory
      } else if (!isAbsolute) {
        normalized.push(segment); // Keep '..' if not absolute and at root
      }
    } else {
      normalized.push(segment);
    }
  }
  
  const result = normalized.join('/');
  return isAbsolute ? `/${ result}` : result;
}

/**
 * Runtime-specific path utilities
 */
export const pathUtils = {
  /**
   * Join paths in a cross-runtime way
   */
  join(...segments: string[]): string {
    const joined = segments
      .filter(Boolean)
      .join('/')
      .replace(/\/+/g, '/') // Replace multiple slashes with single slash
      .replace(/\\/g, '/'); // Normalize backslashes to forward slashes
    
    return normalizePath(joined);
  },

  /**
   * Join paths asynchronously with enhanced Node.js support
   */
  async joinAsync(...segments: string[]): Promise<string> {
    const runtime = detectRuntime();

    // For Node.js, use the real path module when available
    if (runtime.name === 'node') {
      const pathMod = await loadPathModule();
      if (pathMod?.posix) {
        return pathMod.posix.join(...segments);
      }
    }

    // Fallback implementation with normalization
    const joined = segments
      .filter(Boolean)
      .join('/')
      .replace(/\/+/g, '/') // Replace multiple slashes with single slash
      .replace(/\\/g, '/'); // Normalize backslashes to forward slashes
    
    return normalizePath(joined);
  },

  /**
   * Resolve absolute path in a cross-runtime way (synchronous)
   */
  resolve(path: string): string {
    const runtime = detectRuntime();

    // Simple synchronous resolution for testing
    if (path.startsWith('/')) {
      return path; // Already absolute
    }

    if (runtime.name === 'node' && typeof process !== 'undefined' && process.cwd) {
      return normalizePath(`${process.cwd() }/${ path}`);
    }

    if (runtime.name === 'browser' && typeof location !== 'undefined') {
      return new URL(path, location.href).href;
    }

    // Fallback: normalize and return as-is for relative paths
    return normalizePath(path);
  },

  /**
   * Resolve absolute path in a cross-runtime way (asynchronous)
   */
  async resolveAsync(path: string): Promise<string> {
    const runtime = detectRuntime();

    if (runtime.name === 'node') {
      const pathMod = await loadPathModule();
      if (pathMod) {
        return pathMod.resolve(path);
      }
    }

    if (runtime.name === 'browser') {
      return new URL(path, location.href).href;
    }

    // Fallback: normalize and return as-is
    return normalizePath(path);
  },
};

/**
 * Cross-runtime file system operations
 */
export const fsUtils = {
  /**
   * Check if file system operations are available
   */
  isAvailable(): boolean {
    return hasCapability('filesystem');
  },

  /**
   * Read file in a cross-runtime way
   */
  async readFile(path: string): Promise<string> {
    const runtime = detectRuntime();

    if (!runtime.capabilities.filesystem) {
      throw new Error('File system operations not available in this runtime');
    }

    if (runtime.name === 'deno' && Deno) {
      return await Deno.readTextFile(path);
    }

    if (runtime.name === 'bun' && Bun) {
      const file = Bun.file(path);
      return await file.text();
    }

    if (runtime.name === 'node') {
      // Dynamic import to maintain ESM compatibility
      const { readFile } = await import('fs/promises');
      return await readFile(path, 'utf-8');
    }

    throw new Error(`File system operations not implemented for ${runtime.name}`);
  },

  /**
   * Write file in a cross-runtime way
   */
  async writeFile(path: string, content: string): Promise<void> {
    const runtime = detectRuntime();

    if (!runtime.capabilities.filesystem) {
      throw new Error('File system operations not available in this runtime');
    }

    if (runtime.name === 'deno' && Deno) {
      await Deno.writeTextFile(path, content);
      return;
    }

    if (runtime.name === 'bun' && Bun) {
      await Bun.write(path, content);
      return;
    }

    if (runtime.name === 'node') {
      const { writeFile, mkdir } = await import('fs/promises');
      const parts = path.split('/');
      parts.pop();
      const dir = parts.join('/') || '.';
      try {
        await mkdir(dir, { recursive: true });
      } catch {}
      await writeFile(path, content, 'utf-8');
      return;
    }

    throw new Error(`File system operations not implemented for ${runtime.name}`);
  },
};

/**
 * Cross-runtime subprocess execution
 */
export const processUtils = {
  /**
   * Check if subprocess operations are available
   */
  isAvailable(): boolean {
    return hasCapability('subprocess');
  },

  /**
   * Execute command in a cross-runtime way
   */
  async exec(
    command: string,
    args: string[] = []
  ): Promise<{ stdout: string; stderr: string; code: number }> {
    const runtime = detectRuntime();

    if (!runtime.capabilities.subprocess) {
      throw new Error('Subprocess operations not available in this runtime');
    }

    if (runtime.name === 'deno' && Deno) {
      const cmd = new Deno.Command(command, { args });
      const { code, stdout, stderr } = await cmd.output();

      return {
        code,
        stdout: new TextDecoder().decode(stdout),
        stderr: new TextDecoder().decode(stderr),
      };
    }

    if (runtime.name === 'bun' && Bun) {
      const proc = Bun.spawn([command, ...args], {
        stdout: 'pipe',
        stderr: 'pipe',
      });

      const [stdout, stderr] = await Promise.all([
        new Response(proc.stdout).text(),
        new Response(proc.stderr).text(),
      ]);

      await proc.exited;

      return {
        code: proc.exitCode ?? 0,
        stdout,
        stderr,
      };
    }

    if (runtime.name === 'node') {
      const { spawn } = await import('child_process');

      return new Promise((resolve, reject) => {
        const extraPyPath = pathUtils.join(process.cwd(), 'tywrap_ir');
        const env = {
          ...process.env,
          PYTHONPATH: `${extraPyPath}${process.env.PYTHONPATH ? `:${process.env.PYTHONPATH}` : ''}`,
        };
        const child = spawn(command, args, { env });
        let stdout = '';
        let stderr = '';

        child.stdout?.on('data', data => {
          stdout += data.toString();
        });

        child.stderr?.on('data', data => {
          stderr += data.toString();
        });

        child.on('close', code => {
          resolve({ code: code ?? 0, stdout, stderr });
        });

        child.on('error', reject);
      });
    }

    throw new Error(`Subprocess operations not implemented for ${runtime.name}`);
  },
};

/**
 * Cross-runtime hashing and stable stringify utilities
 */
export const hashUtils = {
  async sha256Hex(text: string): Promise<string> {
    const runtime = detectRuntime();
    // Node path
    if (runtime.name === 'node') {
      try {
        const crypto = await import('node:crypto');
        return crypto.createHash('sha256').update(text).digest('hex');
      } catch {
        // fallthrough to web crypto
      }
    }
    // Web/Deno/Bun path via SubtleCrypto
    if (typeof globalThis.crypto?.subtle !== 'undefined') {
      const data = new TextEncoder().encode(text);
      const digest = await globalThis.crypto.subtle.digest('SHA-256', data);
      const bytes = Array.from(new Uint8Array(digest));
      return bytes.map(b => b.toString(16).padStart(2, '0')).join('');
    }
    // Fallback to DJB2 (non-crypto) for unknown runtimes
    let hash = 5381;
    for (let i = 0; i < text.length; i++) {
      hash = (hash << 5) + hash + text.charCodeAt(i);
      hash |= 0;
    }
    return Math.abs(hash).toString(16);
  },
};
