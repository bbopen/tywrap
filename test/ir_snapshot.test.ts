import { describe, it, expect } from 'vitest';
import { processUtils } from '../src/utils/runtime.js';

function sanitizeIr(ir: any) {
  return {
    module: ir.module,
    functionNames: (ir.functions ?? [])
      .map((f: any) => f.name)
      .sort()
      .slice(0, 30),
    classNames: (ir.classes ?? []).map((c: any) => c.name).sort(),
    // Omit metadata and docstrings from snapshot for stability
  };
}

describe('IR golden snapshot - math', () => {
  it('matches sanitized snapshot', async () => {
    const result = await processUtils.exec('python3', [
      '-m',
      'tywrap_ir',
      '--module',
      'math',
      '--no-pretty',
    ]);
    expect(result.code).toBe(0);
    const ir = JSON.parse(result.stdout);
    expect(sanitizeIr(ir)).toMatchSnapshot();
  });
});
