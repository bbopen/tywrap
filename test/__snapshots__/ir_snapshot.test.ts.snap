// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IR golden snapshot - math > matches sanitized snapshot 1`] = `
{
  "classNames": [],
  "functionNames": [
    "acos",
    "acosh",
    "asin",
    "asinh",
    "atan",
    "atan2",
    "atanh",
    "cbrt",
    "ceil",
    "comb",
    "copysign",
    "cos",
    "cosh",
    "degrees",
    "dist",
    "erf",
    "erfc",
    "exp",
    "exp2",
    "expm1",
    "fabs",
    "factorial",
    "floor",
    "fma",
    "fmod",
    "frexp",
    "fsum",
    "gamma",
    "gcd",
    "isclose",
  ],
  "module": "math",
}
`;
